sql, server: add api endpoint for replica matrix

vilterp opened the pull request:
This endpoint will be used by the Admin UI (prototype here: #20500). It shows how many replicas each node has for each table, and returns zone configs so a user can see if replicas are being allocated correctly.

UI PR on top of this: https://github.com/vilterp/cockroach/pull/2

Commit:
server: add endpoint to get # replicas for each (node, table) pair

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22344)
<!-- Reviewable:end -->

vilterp commented on pkg/server/admin_test.go:
```diff
@@ -1369,3 +1369,53 @@ func TestAdminAPIFullRangeLog(t *testing.T) {
+	}
+
+	if len(seenCells) < len(expectedCells) {
+		t.Fatalf("expected %d cells, only saw %v. Response: %v", len(expectedCells), seenCells, resp)
```
this passes locally, but not in CI. Not sure why — it sees `roachblog.posts`, but not `roachblog.comments`.

BramGruneir commented:
*[pkg/server/admin_test.go, line 1414 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L4Ns00kNmFk5z1aRbcx-r2-1414:-L5PByjwMuU3hVduVcCi:bbn4uxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/admin_test.go#L1414)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

this passes locally, but not in CI. Not sure why — it sees `roachblog.posts`, but not `roachblog.comments`.
</blockquote></details>

I'm guessing it's a timing issue.  
Maybe add a loop that looks for the existence of the table?  If that still fails, we have a bigger issue on our hands.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Reviewed 2 of 3 files at r1, 5 of 6 files at r2.
Review status: 7 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 1316 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L5PFBQw_Jfd86e_-0wJ:-L5PFBQw_Jfd86e_-0wK:b-c97udq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/admin.go#L1316)):*
> ```Go
> }
> 
> func (s *adminServer) ReplicaMatrix(
> ```

This is going to need a comment.

---

*[pkg/server/admin_test.go, line 1394 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L5PFs70jty51A1nQfes:-L5PFs70jty51A1nQfet:b8ov1rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/admin_test.go#L1394)):*
> ```Go
> 		t.Fatal(err)
> 	}
> ```

Can you check the zone_config stuff here too?  There are some changes coming to them and it would probably be a good idea to make sure this doesn't rot when that happens.

---

*[pkg/server/serverpb/admin.proto, line 505 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L5PG6QU0OElTQ89btlE:-L5PG6QU0OElTQ89btlF:bb7gcla) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/serverpb/admin.proto#L505)):*
> ```protobuf
> 
> message ReplicaMatrixRequest {
>   int64 foo = 1;
> ```

foo?

---

*[pkg/server/serverpb/admin.proto, line 512 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L5PKY7vlqeunfLN-8Uc:-L5PKY7xhMcs7mVVYsWB:b-yb0rqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/serverpb/admin.proto#L512)):*
> ```protobuf
>     // id is 0 for the default zone config.
>     // Otherwise, it's the id of a database or table descriptor.
>     int64 id = 1;
> ```

Why not split this up.  DatabaseID, TableID and also partitionID (or however the new partition works)?  But even more than that, why have this at all if you have cli_specifier?

---

*[pkg/sql/crdb_internal.go, line 1492 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L5PSpVxysOhmr5qtKsH:-L5PSpVxysOhmr5qtKsI:bmn08zq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/sql/crdb_internal.go#L1492)):*
> ```Go
> }
> 
> var crdbInternalReplicasTable = virtualSchemaTable{
> ```

This should be looked at by @knz.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Finally passing under stress. PTAL!

---

Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/server/admin.go, line 1316 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L5PFBQw_Jfd86e_-0wJ:-L63I9aVwYIRRL7tdfQ-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/admin.go#L1316)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This is going to need a comment.
</blockquote></details>

Done.

---

*[pkg/server/admin_test.go, line 1394 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L5PFs70jty51A1nQfes:-L63IEcjXucfMtAC90Do:beqxmjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/admin_test.go#L1394)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you check the zone_config stuff here too?  There are some changes coming to them and it would probably be a good idea to make sure this doesn't rot when that happens.
</blockquote></details>

Checked below, but only minimally. UI currently just dumps the response; don't want to make to strict a test and have it fail when zone configs change.

---

*[pkg/server/admin_test.go, line 1414 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L4Ns00kNmFk5z1aRbcx-r2-1414:-L63IPu3BkYl2RLpsm5H:b68obmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/admin_test.go#L1414)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I'm guessing it's a timing issue.  
Maybe add a loop that looks for the existence of the table?  If that still fails, we have a bigger issue on our hands.
</blockquote></details>

Finally fixed.

---

*[pkg/server/serverpb/admin.proto, line 505 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L5PG6QU0OElTQ89btlE:-L63ITpy3t8Mo_STtlhs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/serverpb/admin.proto#L505)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

foo?
</blockquote></details>

Done.

---

*[pkg/server/serverpb/admin.proto, line 512 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L5PKY7vlqeunfLN-8Uc:-L63IWLCVq4RPZhaZV3B:bx92qaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/serverpb/admin.proto#L512)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Why not split this up.  DatabaseID, TableID and also partitionID (or however the new partition works)?  But even more than that, why have this at all if you have cli_specifier?
</blockquote></details>

Just going to rely on CLI specifier for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-:-L63IeqjTjzAwR8JdiAK:beswjdf)*
<!-- Sent from Reviewable.io -->

knz commented:
Wowzer!

"Slow for large clusters" - you don't say!

scan from keys.MinKey to keys.MaxKey, no less. And the result computed with a left outer join on the ranges table, even better. What's next, a `crdb_internal.all_data` vtable that scans MinKey to MaxKey on the entire data space, and then a left outer join on `crdb_internal.all_ranges` to see which data belongs to which range? I see the PR coming, "Slow for large clusters, but might be useful in the admin UI".

I'm only half joking.

Look, I don't care that "it seems to work". The time/memory costs of this thing are just out of control. As long as we prepopulate the vtables in memory and any query containing a vtable is not distributed, this is just not good.

The way to do this needs a bit more work.

1) you need to have the UI screen paginate (obviously). For each "current page" determine the node bounds or the range bounds (Depending on which page). Do some UI design work so that there's a notion of "how much data to show on this page", so that it makes sense to only show a subset of the data by default.

2) make the admin RPC endpoint take bound parameters that limit how much data is queried.

3) write the loop that processes the result of the scan inside the admin RPC endpoint. Don't go to SQL. Write the predicate that filters out rows *based on the bounds given in the RPC parameters* inside the loop.

4)  Also, inside that loop, do your "left join on the ranges table" - that's merely checking the key prefixes for each replica, check which table ID is present in there, and look up the table descriptor to get a table/database name.   Ensure these descriptors are cached between lookups.

---

Reviewed 3 of 9 files at r3, 5 of 6 files at r4.
Review status: 8 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-:-L66gbOZjLvz8zjXt5fk:brl0pks)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 8 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/crdb_internal.go, line 1498 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L66izP640Qg-Ag77bS7:-L66izP640Qg-Ag77bS8:b-9ldadc) ([raw file](https://github.com/cockroachdb/cockroach/blob/addcadffd06caa18fda9ce72b10713951473e247/pkg/sql/crdb_internal.go#L1498)):*
> ```Go
> var crdbInternalReplicasTable = virtualSchemaTable{
> 	schema: `
> CREATE TABLE crdb_internal.replicas (
> ```

I don't support the addition of this vtable (See my top-level comment)

But in any case, any new vtable here needs a logic test in `testdata/logic_tests/crdb_internal`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344)*
<!-- Sent from Reviewable.io -->

knz commented:
Not to mention that your current implementation uses the internal executor and thus doesn't do any memory accounting. It's not just "slow for large clusters" - it would promply OOM kill the node with sufficiently many replicas.

---

Review status: 8 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-:-L66jONHchirG6FNEMkF:b-2a1e5q)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Hey Raphael, this doesn't scan the entire key range, it scans the entire meta2 range (look at `scanMetaKVs`). This is still a large scan, but not as bad as you're suggesting. I realize it will still allocate a lot of datums, so I can remove this vtable and rewrite the endpoint to accumulate a per-table per-node account directly.

With an implementation like that, would it be possible to get away without either paginating the UI or maintaining some kind of incremental materialized view of how many replicas there are on each node, for each table?

In general I prefer to write admin UI endpoints against crdb_internal, since it gives a nice separation of concerns between internal data structures and external data needs. I realize that pushing query operations down into vtable implementations is a bigger project though.

Also, why does the internal executor not do memory accounting? There are a lot of calls to it in this file; can we do something to add memory accounting to all of them?

knz commented:
Three separate concerns:

- single loop, embed the descriptor lookup without join, then avoid pagination. My opinion: "maybe?" What about you make an estimate of what it would cost on a db with a moderate number of ranges (say, 300.000 ranges) and then we talk.

- using crdb_internal vs custom code. Yeah I like `crdb_internal` too but this currently forces the data to be populated in memory upfront. I think that you should shy away from (ab)using this abstraction until we have a good story to turn the vtable constructors into proper generators (that can produce their data in a streaming fashion) - see #23052 to follow up on that.

- re. memory accounting for SQL used by admin RPC endpoints. This is a major limitation which doesn't have a tracking issue yet. Mind creating it and explain the problem at hand?

Commit:
server: add benchmark for ReplicaMatrix endpoint

Benchstat output: (count=10)

name                     time/op
AdminAPIReplicaMatrix-8  385ms ±84%

Benchmark count -10 output:

goos: darwin
goarch: amd64
pkg: github.com/cockroachdb/cockroach/pkg/server
BenchmarkAdminAPIReplicaMatrix-8   	      10	 168937559 ns/op
BenchmarkAdminAPIReplicaMatrix-8   	      10	 432137334 ns/op
BenchmarkAdminAPIReplicaMatrix-8   	      10	 168379098 ns/op
BenchmarkAdminAPIReplicaMatrix-8   	      10	 177149986 ns/op
BenchmarkAdminAPIReplicaMatrix-8   	       5	 206646397 ns/op
BenchmarkAdminAPIReplicaMatrix-8   	      10	 182270207 ns/op
BenchmarkAdminAPIReplicaMatrix-8   	      10	 170547378 ns/op
BenchmarkAdminAPIReplicaMatrix-8   	       5	 204531985 ns/op
BenchmarkAdminAPIReplicaMatrix-8   	      10	 172730136 ns/op
BenchmarkAdminAPIReplicaMatrix-8   	      10	 165367664 ns/op
PASS
ok  	github.com/cockroachdb/cockroach/pkg/server	113.757s

Release note: None

vilterp commented:
Responding to your concerns:

1. Rewrote the endpoint to use a loop. Rough estimate: logging out the size of the KV pairs that are coming back from the meta2 scan shows that they average about 40 bytes. 40 bytes * 300k = 12 mb. This is a little big, but not crazy. I think it's within the range of what time series queries often allocate (cc @mrtracy).<br/><br/>Would be good to get real numbers; not sure how to create that many ranges though. Is there a test that does this which I can copy?
2. That's fair. Looking forward to improved crdb_internal.
3. It looks like these endpoints actually don't use `InternalExecutor`, and do track their memory usage. The SQL is executed by `s.server.sqlExecutor`, which looks like it's the server's top-level `sql.Executor`. Also, it looks like each endpoint which uses SQL calls `NewContextAndSessionForRPC`, which allocates session with a new memory monitor and passes it to `ExecuteStatementsBuffered`.

knz commented:
Regarding point 3 TIL. Thanks for checking!

knz commented:
Reviewed 1 of 6 files at r4, 6 of 8 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 1364 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L6LvYsQtlSWm1-aSSxg:-L6LvYsRxPgd-Cs_lsos:byywaun) ([raw file](https://github.com/cockroachdb/cockroach/blob/48461c41f3f5dcccb0bc9d089e3f9c45218aee65/pkg/server/admin.go#L1364)):*
> ```Go
> 	databaseIDForTableID := map[uint64]uint64{}
> 
> 	tablesQuery := `SELECT name, table_id, database_name, parent_id FROM crdb_internal.tables`
> ```

1) Add a comment: "This relies on crdb_internal.tables returning data even for newly added tables and deleted tables (as opposed to e.g. information_schema) because we are interested in the data for all ranges, not just visible ranges"

2) make it clearer that you want all tables by querying with `FROM "".crdb_internal.tables` (currently no functional difference, but that might change)

3) given that you're re-building essentially the same data structures as what `crdb_internal.tables` does, below, I wonder if this is the right approach. Please consider exporting `internalLookupCtx` (sql/resolver.go) instead and reuse that here:

```go
descs, err := sql.GetAllDescriptors(...)
if err != nil { ... }
lookupCtx := sql.NewInternalLookupCtx(descs, nil)
// use lookupCtx below
```

---

*[pkg/server/admin.go, line 1395 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344#-L6LwAznphTPEq6ZRXfv:-L6LwAznphTPEq6ZRXfw:b-vwa6ma) ([raw file](https://github.com/cockroachdb/cockroach/blob/48461c41f3f5dcccb0bc9d089e3f9c45218aee65/pkg/server/admin.go#L1395)):*
> ```Go
> 		// Get zone config for table.
> 		zoneConfigQuery := fmt.Sprintf(
> 			`SELECT id, cli_specifier FROM [EXPERIMENTAL SHOW ZONE CONFIGURATION FOR TABLE %s.%s]`, dbName, tableName,
> ```

Not correct if the db/table names contain special characters.

I recommend `dbName := (*string)(row[2].(*tree.DString))` above then `(*tree.Name)(dbName)` here (rendering a tree.Name via %s does the right thing).

Also add a test with special names to check.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22344)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Comments addressed and/or responded to; PTAL. 

---

Reviewed 1 of 6 files at r4, 3 of 8 files at r5, 5 of 6 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/server/admin.go, line 1364 at r7](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L6LvYsQtlSWm1-aSSxg:-L9RkawsE9w4C2jqzeQQ:b1ificn) ([raw file](https://github.com/cockroachdb/cockroach/blob/48461c41f3f5dcccb0bc9d089e3f9c45218aee65/pkg/server/admin.go#L1364)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) Add a comment: "This relies on crdb_internal.tables returning data even for newly added tables and deleted tables (as opposed to e.g. information_schema) because we are interested in the data for all ranges, not just visible ranges"

2) make it clearer that you want all tables by querying with `FROM "".crdb_internal.tables` (currently no functional difference, but that might change)

3) given that you're re-building essentially the same data structures as what `crdb_internal.tables` does, below, I wonder if this is the right approach. Please consider exporting `internalLookupCtx` (sql/resolver.go) instead and reuse that here:

```go
descs, err := sql.GetAllDescriptors(...)
if err != nil { ... }
lookupCtx := sql.NewInternalLookupCtx(descs, nil)
// use lookupCtx below
```

</blockquote></details>

1 and 2: Done.

3: I'd prefer to keep `internalLookupCtx` internal and use the SQL interface, although I realize it creates some extra GC load by allocating a datum for each table and throwing it away. The virtual table implementation doesn't use `internalLookupCtx`; what does it provide which is needed here?

---

*[pkg/server/admin.go, line 1395 at r7](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L6LwAznphTPEq6ZRXfv:-L9Rlk9V4WM-cNMzgk4T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/48461c41f3f5dcccb0bc9d089e3f9c45218aee65/pkg/server/admin.go#L1395)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Not correct if the db/table names contain special characters.

I recommend `dbName := (*string)(row[2].(*tree.DString))` above then `(*tree.Name)(dbName)` here (rendering a tree.Name via %s does the right thing).

Also add a test with special names to check.
</blockquote></details>

Done.

---

*[pkg/sql/crdb_internal.go, line 1498 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L66izP640Qg-Ag77bS7:-L9Rlq6V0v_eEBBiIAiF:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/addcadffd06caa18fda9ce72b10713951473e247/pkg/sql/crdb_internal.go#L1498)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't support the addition of this vtable (See my top-level comment)

But in any case, any new vtable here needs a logic test in `testdata/logic_tests/crdb_internal`
</blockquote></details>

Removed.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-:-L9RlvDk9ozN8LBwQZ4h:bb6rpi6)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Sigh, test failing because table descriptor IDs are sometimes different. `make stress` couldn't repro this locally over 135 runs.

vilterp commented on pkg/server/serverpb/admin.proto:
```diff
@@ -514,6 +514,40 @@ message QueryPlanResponse {
+    map<int32, int64> replica_count_by_node_id = 2
+      [(gogoproto.castkey) = "github.com/cockroachdb/cockroach/pkg/roachpb.NodeID"];
+
+    int64 zone_config_id = 3;
```
@benesch this is the PR I mentioned IRL today. Partitions would slot in here, under tables. PTAL at this PR when you can; would like to get it and the accompanying frontend PR in soon, now that we're out of code freeze!

benesch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I only reviewed from the perspective of future partitioning extensibility.

---

Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/server/serverpb/admin.proto, line 512 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L5PKY7vlqeunfLN-8Uc:-L9sBKGF56pkiSpJpf-9:bsn1wc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/serverpb/admin.proto#L512)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Just going to rely on CLI specifier for now.
</blockquote></details>

CLI specifiers need to go away, but that won't be for a long time. (Totally fine to use them for now.)

---

*[pkg/server/serverpb/admin.proto, line 533 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L9sAqmx97wXL-mHKKDq:-L9sAqmx97wXL-mHKKDr:b-kdm95n) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d260d2b621daa65f5f59a83317a5962ee1666f0/pkg/server/serverpb/admin.proto#L533)):*
> ```protobuf
> 
>   message TableInfo {
>     map<int32, int64> replica_count_by_node_id = 2
> ```

optional nit: you skipped 1!

---

*[pkg/server/serverpb/admin.proto, line 536 at r11](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L4Ns00kNmFk5z1aRbd--r11-536:-L9sASMy-wgdgt6iqUyV:blkj5j8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cae00485c902171d8481f033f8c9ac194383531c/pkg/server/serverpb/admin.proto#L536)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

@benesch this is the PR I mentioned IRL today. Partitions would slot in here, under tables. Curious how you think that should look, exactly — I'll note that I'd prefer to merge this basic version soon and add partitions and indexes in another PR, even if it breaks backwards-compatibility on this API. I think that's ok this early in the cycle, esp. since the UI is the only consumer.

Please take a look at this PR when you can; would like to get it and the accompanying frontend PR in soon, now that we're out of code freeze!
</blockquote></details>

This makes sense to me. Here's what I'd recommend for partitions:

```
map<string, PartitionInfo partition_info = 4

// ...

message PartitionInfo {
  map<int32, int64> replica_count_by_node_id = 1;
  int64 zone_config_id = 2;
}
```

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-:-L9sC-QP78earDNuD9CQ:b-1kt37a)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/server/serverpb/admin.proto, line 512 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L5PKY7vlqeunfLN-8Uc:-L9sMq_J2IdDPj03h5BK:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b264646257a39c187af995186e8907ffb933235/pkg/server/serverpb/admin.proto#L512)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Ok. `TableInfo`s have zone config ids here (obtained with `EXPERIMENTAL SHOW ZONE CONFIGURATION FOR TABLE` in `admin.go`), so we're not actually using the CLI specifier to establish the linkage.
</blockquote></details>

Ack.

---

*[pkg/server/serverpb/admin.proto, line 536 at r11](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L4Ns00kNmFk5z1aRbd--r11-536:-L9sMj4P48GylLsFtVHC:by81g2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cae00485c902171d8481f033f8c9ac194383531c/pkg/server/serverpb/admin.proto#L536)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Cool. Wouldn't I need `IndexInfo` between `TableInfo` and `PartitionInfo` though?
</blockquote></details>

Not unless you want it for display in the UI. Partition names are unique across all indices within a table.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Fixed the nit.

@knz, do you think this is ready to merge? You had mentioned adding a parameter to narrow the request down to a specific table or database — I think that's a good idea but I'll leave it for a later PR; it doesn't require any structural changes here.

---

Reviewed 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/server/serverpb/admin.proto, line 533 at r10](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L9sAqmx97wXL-mHKKDq:-L9u_GZZ-tniCWpnNjcg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d260d2b621daa65f5f59a83317a5962ee1666f0/pkg/server/serverpb/admin.proto#L533)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

optional nit: you skipped 1!
</blockquote></details>

Done.

---

*[pkg/server/serverpb/admin.proto, line 536 at r11](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L4Ns00kNmFk5z1aRbd--r11-536:-L9u_I3SEiGcg0MtQMYm:b-uxtbzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/cae00485c902171d8481f033f8c9ac194383531c/pkg/server/serverpb/admin.proto#L536)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Not unless you want it for display in the UI. Partition names are unique across all indices within a table.
</blockquote></details>

Ok, good to know they're unique. Still seems like you would want to see the the partitions broken out by index, not in one giant list, right? How does it look in a `CREATE TABLE` if you partition multiple indices?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-:-L9uaawQ4iFJ87gdSh0K:b-gegra0)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/server/serverpb/admin.proto, line 536 at r11](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L4Ns00kNmFk5z1aRbd--r11-536:-L9umDmf9wroJn5u5txu:b-srn8ps) ([raw file](https://github.com/cockroachdb/cockroach/blob/cae00485c902171d8481f033f8c9ac194383531c/pkg/server/serverpb/admin.proto#L536)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Ok, good to know they're unique. Still seems like you would want to see the the partitions broken out by index, not in one giant list, right? How does it look in a `CREATE TABLE` if you partition multiple indices?
</blockquote></details>

Yeah, I think you're right. A table with multiple partitioned indices looks like this:

```
CREATE TABLE t (
  a INT PRIMARY KEY,
  b INT,
  c INT,
  INDEX (b) PARTITION BY...
  INDEX (b, c) PARTITION BY...
) PARTITION BY ...
```

So even though the partition namespace is shared across all indices, you do declare them within an index.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I'd love to see more special characters exercised in the tests.

---

Reviewed 3 of 6 files at r8, 1 of 1 files at r11, 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/server/admin_test.go, line 1333 at r12](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-L9yqw-l4cmyRLATLSTT:-L9yqw-l4cmyRLATLSTU:b-6qe3os) ([raw file](https://github.com/cockroachdb/cockroach/blob/08c2d08fda1348025a4968aa43919e30f76e05e7/pkg/server/admin_test.go#L1333)):*
> ```Go
> 	)`)
> 	// Test special characters in DB and table names.
> 	sqlDB.Exec(t, `CREATE DATABASE "spec'chars"`)
> ```

A single quote is not a sufficiently special character for an identifier. (I'm nearly sure it would work fine with or without your change after my comment)

Try identifiers containing backslashes and double quotes.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22344#-:-L9yrNNjCBiyXRh6Q5Dw:b-o976lp)*
<!-- Sent from Reviewable.io -->

vilterp commented:
TFTRs! Added more special characters to the test, @knz.

bors r+

vilterp commented:
Hm, looks like it failed under stress, though it has succeeded in previous build runs. I don't really know what to do to deflake, other than give it a longer timeout — the test creates a table, and then asserts that it rebalances to all three nodes in the test cluster.

Will keep an eye on this and revert if it's flaky on master.

andreimatei commented:
The first commit in this PR is from Dec 4th. No wonder I was so confused about why I'm getting rebase conflicts. Congrats on the perseverance!

vilterp commented:
Hah yeah, funny that it doesn't update the date when you rebase / squash. UI still isn't checked in; also first pushed 12/4…
